trinity/protocol/eth/managers.py

```diff
@@ -57,38 +67,47 @@
 ]
 
 
-class GetBlockHeadersRequestManager(BaseGetBlockHeadersRequestManager):
+class HeaderRequestArgs(NamedTuple):
+    block_number_or_hash: BlockIdentifier
+    max_headers: int
+    skip: int
+    reverse: bool
+
+
+BlockHeadersRequestManager = RequestManager['ETHPeer', Tuple[BlockHeader, ...]]
+
+
+class GetBlockHeadersRequestManager(BaseService):
     msg_queue_maxsize = 100
 
     _response_msg_type: Type[Command] = BlockHeaders
 
-    # All `RequestManager` classes are expected to implement the `__call__`
-    # method, including changing the function signature, thus the
-    # `# type: ignore` here is both expected and required.
-    async def __call__(self,  # type: ignore
+    def __init__(self, peer: 'ETHPeer', token: CancelToken) -> None:
+        super().__init__(token)
+        self._peer = peer
+        self._request_manager = BlockHeadersRequestManager(peer, BlockHeaders, token)
+
+    async def __call__(self,
                        block_number_or_hash: BlockIdentifier,
                        max_headers: int = None,
                        skip: int = 0,
                        reverse: bool = True,
                        timeout: int = None) -> Tuple[BlockHeader, ...]:
-        request = HeaderRequest(
-            block_number_or_hash,
-            max_headers,
-            skip,
-            reverse,
-        )
-        return await self._request_and_wait(request, timeout)
 
-    def _send_sub_proto_request(self, request: HeaderRequest) -> None:
-        self._peer.sub_proto.send_get_block_headers(request)
+        request_args = HeaderRequestArgs(block_number_or_hash, max_headers, skip, reverse)
+        make_request = partial(self._peer.sub_proto.send_get_block_headers, *request_args)
+        validator = HeaderRequest(*request_args)
+        manager = self._request_manager
 
-    async def _normalize_response(self,
-                                  msg: Tuple[BlockHeader, ...],
-                                  ) -> Tuple[BlockHeader, ...]:
-        return msg
+        async for message in manager.message_candidates(make_request, self.logger, timeout):
+            with manager.complete_if_valid(len(message)):
+                validator.validate_result(message)
+                return message
 
-    def _get_item_count(self, msg: Tuple[BlockHeader, ...]) -> int:
-        return len(msg)
+        raise ValidationError("Manager is not pending a response, but no valid response received")
+
+    async def _run(self) -> None:
+        await self._request_manager.run()
```

Comment:
It'd be nice if we could get rid of the extra service layer here.  See my other comment about removing this class entirely for how we might do that.